File: 846.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        if len(hand) % groupSize != 0: 
            return False
        
        freq = Counter(hand)
        hand.sort()

        for card in hand:
            if freq[card] == 0:
                continue
            
            for i in range(groupSize):
                nextCard = card + i
                if freq[nextCard] == 0:
                    return False
                freq[nextCard] = freq[nextCard] - 1
        
        return True

File: 846.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        if(hand.length % groupSize != 0) return false;

        Map<Integer, Integer> map = new HashMap<>();
        for(int card : hand) {
            map.put(card, map.getOrDefault(card, 0) + 1);
        }

        Arrays.sort(hand);

        for(int card : hand) {
            if(map.get(card) == 0) continue;

            for(int i = 0; i < groupSize; i++){
                int next = card + i;
                if(map.getOrDefault(next, 0) == 0) {
                    return false;
                }
                map.put(next, map.get(next) - 1);
            }
        }
        return true;
    }
}
